### PR TITLE
Add parsing to make cd~ a valid command.

### PR DESCRIPTION
**Pull request by GigaScratch**
# PR Summary
`cd..` and `cd\\` currently automatically parse to `Set-Location ..` and `Set-Location \\` respectively. This PR allows the same behavior for `cd~`.
image

## PR Context
Resolves issue #17140 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/9329
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by SteveL-MSFT — DISMISSED** [review]

**Comment by iSazonov** [comment]
Should we document this?

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 1]
```diff
@@ -4709,6 +4709,7 @@ internal static SessionStateAliasEntry[] BuiltInAliases
            // Functions that don't require full language mode
             SessionStateFunctionEntry.GetDelayParsedFunctionEntry("cd..", "Set-Location ..", isProductCode: true, languageMode: systemLanguageMode),
             SessionStateFunctionEntry.GetDelayParsedFunctionEntry("cd\\", "Set-Location \\", isProductCode: true, languageMode: systemLanguageMode),
+            SessionStateFunctionEntry.GetDelayParsedFunctionEntry("cd~", "Set-Location ~", isProductCode: true, languageMode: systemLanguageMode),
```
[the remark this answers is not in the document]
Comment (GigaScratch) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4712]: Added Test for `cd~` case and `cd..` case. I couldn't get network drives to mock well so skipped `cd\\`

File: test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1 [hunk 2]
```diff
@@ -230,4 +230,19 @@ Describe "Set-Location" -Tags "CI" {
...
+        It 'Should go to Filesystem home on cd~ run' {
+            Set-Location 'TestDrive:\'
+            cd~
+            (Get-Location).Path | Should -BeExactly (Get-PSProvider FileSystem).Home
+        }
+        It 'Should go to the Parent folder on cd.. run'{
+            $ParentDir = 'TestDrive:\'
+            New-Item -Path 'TestDrive:\' -Name 'Directory1' -ItemType Directory
+            Set-Location 'TestDrive:\Directory1'
+            cd..
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:244]: Have we a test for `cd\`? Please add if no.

**Review by iSazonov — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1 [hunk 3]
```diff
@@ -230,4 +230,36 @@ Describe "Set-Location" -Tags "CI" {
...
+            Set-Location 'TestDrive:\Directory1\Directory2'
+            cd\
+            (Get-Location).Path | Should -BeExactly $DriveRoot
+        }
+        It 'Should go to root of current drive on cd\ run (Linux/Mac)' -Skip:($IsWindows){
+            Set-Location 'TestDrive:\'
+            New-Item -Path 'TestDrive:\Directory1' -Name 'Directory2' -ItemType Directory
+            Set-Location 'TestDrive:\Directory1\Directory2'
+            cd\
+            (Get-Location).Path | Should -BeExactly "/"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:262]: nit: his difference with Windows looks like a bug.

**Review by iSazonov — APPROVED** [review]
LGTM with one comment.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1 (hunk 3 above)
Comment (GigaScratch) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:262]: #18325 Created

**Comment by GigaScratch** [comment]
Can you please associate the development of this with the issue #17140 so I can get my free internet points for Hacktoberfest? thanks.

**Comment by iSazonov** [comment]
@sdwheeler Should we document this?

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Tests were added. Please take another look when you have a chance.

**Review by SteveL-MSFT — APPROVED** [review]

**Pull request merged**